File: Character/Services/Big5AnalysisService.swift
import Foundation
import FirebaseFirestore
import FirebaseFunctions

class Big5AnalysisService: ObservableObject {
    private let db = Firestore.firestore()
    private let cache = Big5AnalysisCache.shared
    private let subscriptionManager = SubscriptionManager.shared
    
    @Published var currentAnalysisData: Big5AnalysisData?
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    // MARK: - PersonalityKey Generation
    
    func generatePersonalityKey(scores: Big5Scores, gender: String) -> String {
        let o = roundToFiveScale(scores.openness)
        let c = roundToFiveScale(scores.conscientiousness)
        let e = roundToFiveScale(scores.extraversion)
        let a = roundToFiveScale(scores.agreeableness)
        let n = roundToFiveScale(scores.neuroticism)
        
        return "O\(o)_C\(c)_E\(e)_A\(a)_N\(n)_\(gender)"
    }
    
    private func roundToFiveScale(_ score: Double) -> Int {
        return max(1, min(5, Int(round(score))))
    }
    
    // MARK: - Analysis Data Fetching
    
    func fetchAnalysisData(personalityKey: String, completion: @escaping (Result<Big5AnalysisData, Error>) -> Void) {

        // キャッシュから取得を試行
        if let cachedData = cache.getCachedAnalysis(personalityKey: personalityKey) {
            DispatchQueue.main.async {
                self.currentAnalysisData = cachedData
                completion(.success(cachedData))
            }
            return
        }

        isLoading = true
        errorMessage = nil

        db.collection("Big5Analysis").document(personalityKey).getDocument { [weak self] document, error in
            if let error = error {
                DispatchQueue.main.async {
                    self?.isLoading = false
                    self?.errorMessage = error.localizedDescription
                    completion(.failure(error))
                }
                return
            }

            if let document = document, document.exists, let data = document.data() {
                // データが存在する場合は既存の処理
                do {
                    let analysisData = try self?.parseAnalysisData(from: data, personalityKey: personalityKey)
                    if let analysisData = analysisData {
                        DispatchQueue.main.async {
                            self?.isLoading = false
                            self?.currentAnalysisData = analysisData
                            self?.cache.cacheAnalysis(key: personalityKey, data: analysisData)
                            completion(.success(analysisData))
                        }
                    }
                } catch {
                    DispatchQueue.main.async {
                        self?.isLoading = false
                        self?.errorMessage = "データの解析に失敗しました"
                        completion(.failure(error))
                    }
                }
            } else {
                // データが存在しない場合はエラーを返す（画面では生成しない）
                let notFoundError = NSError(domain: "Big5AnalysisService", code: 404, userInfo: [NSLocalizedDescriptionKey: "解析データがまだ生成されていません"])
                DispatchQueue.main.async {
                    self?.isLoading = false
                    self?.errorMessage = "解析データがまだ生成されていません"
                    completion(.failure(notFoundError))
                }
            }
        }
    }
    
    // MARK: - Data Parsing
    
    private func parseAnalysisData(from data: [String: Any], personalityKey: String) throws -> Big5AnalysisData {
        let lastUpdated = (data["last_updated"] as? Timestamp)?.dateValue() ?? Date()
        
        let analysis20 = parseAnalysisLevel(from: data, level: "analysis_20")
        let analysis50 = parseAnalysisLevel(from: data, level: "analysis_50")
        let analysis100 = parseAnalysisLevel(from: data, level: "analysis_100")
        
        return Big5AnalysisData(
            personalityKey: personalityKey,
            lastUpdated: lastUpdated,
            analysis20: analysis20,
            analysis50: analysis50,
            analysis100: analysis100
        )
    }
    
    private func parseAnalysisLevel(from data: [String: Any], level: String) -> [Big5AnalysisCategory: Big5DetailedAnalysis]? {
        guard let levelData = data[level] as? [String: Any] else {
            return nil
        }


        var result: [Big5AnalysisCategory: Big5DetailedAnalysis] = [:]
        let analysisLevel: Big5AnalysisLevel

        switch level {
        case "analysis_20":
            analysisLevel = .basic
        case "analysis_50":
            analysisLevel = .detailed
        case "analysis_100":
            analysisLevel = .complete
        default:
            return nil
        }

        for category in Big5AnalysisCategory.allCases {

            if let categoryData = levelData[category.rawValue] as? [String: Any] {

                if let personalityType = categoryData["personality_type"] as? String,
                   let detailedText = categoryData["detailed_text"] as? String,
                   let keyPoints = categoryData["key_points"] as? [String] {

                    let analysis = Big5DetailedAnalysis(
                        category: category,
                        personalityType: personalityType,
                        detailedText: detailedText,
                        keyPoints: keyPoints,
                        analysisLevel: analysisLevel
                    )

                    result[category] = analysis
                } else {
                }
            } else {
            }
        }

        return result.isEmpty ? nil : result
    }
    
    // MARK: - Character Analysis Data Fetching
    
    func fetchCharacterAnalysis(characterId: String, userId: String, completion: @escaping (Result<Big5AnalysisData, Error>) -> Void) {

        // ユーザーのキャラクター詳細からpersonalityKeyを取得
        db.collection("users").document(userId)
            .collection("characters").document(characterId)
            .collection("details").document("current").getDocument { [weak self] document, error in
                if let error = error {
                    DispatchQueue.main.async {
                        self?.errorMessage = "キャラクター詳細の取得に失敗: \(error.localizedDescription)"
                        completion(.failure(error))
                    }
                    return
                }

                guard let data = document?.data() else {
                    let notFoundError = NSError(domain: "Big5AnalysisService", code: 404, userInfo: [NSLocalizedDescriptionKey: "キャラクター詳細が見つかりませんでした"])
                    DispatchQueue.main.async {
                        self?.errorMessage = "キャラクター詳細が見つかりませんでした"
                        completion(.failure(notFoundError))
                    }
                    return
                }


                guard let personalityKey = data["personalityKey"] as? String else {
                    let notFoundError = NSError(domain: "Big5AnalysisService", code: 404, userInfo: [NSLocalizedDescriptionKey: "キャラクターのpersonalityKeyが見つかりませんでした"])
                    DispatchQueue.main.async {
                        self?.errorMessage = "personalityKeyが設定されていません"
                        completion(.failure(notFoundError))
                    }
                    return
                }


                // personalityKeyで解析データを取得
                self?.fetchAnalysisData(personalityKey: personalityKey, completion: completion)
            }
    }
    
    // MARK: - Analysis Level Determination
    
    func determineAnalysisLevel(answeredCount: Int) -> Big5AnalysisLevel? {
        if answeredCount >= 100 {
            return .complete
        } else if answeredCount >= 50 {
            return .detailed
        } else if answeredCount >= 20 {
            return .basic
        } else {
            return nil
        }
    }
    
    // MARK: - Available Categories for Level
    
    func getAvailableCategories(for level: Big5AnalysisLevel) -> [Big5AnalysisCategory] {
        switch level {
        case .basic:
            // 基本解析では3つのカテゴリーのみ
            return [.career, .romance, .stress]
        case .detailed, .complete:
            // 詳細・完全解析では全5つのカテゴリー
            return Big5AnalysisCategory.allCases
        }
    }
    
    // MARK: - Dynamic Analysis Data Generation
    
    private func generateAnalysisData(personalityKey: String, completion: @escaping (Result<Big5AnalysisData, Error>) -> Void) {

        // Cloud Functionを呼び出してAIで解析データを生成
        let functions = Functions.functions(region: "asia-northeast1")

        // isPremiumフラグでFirebase Functions側でモデル選択
        // 無料: GPT-4o-mini, 有料: GPT-4o-2024-11-20
        Task { @MainActor in
            let isPremiumValue = subscriptionManager.isPremium

            functions.httpsCallable("generateBig5Analysis").call([
                "personalityKey": personalityKey,
                "isPremium": isPremiumValue
            ]) { [weak self] result, error in
            if let error = error {
                DispatchQueue.main.async {
                    self?.isLoading = false
                    self?.errorMessage = "解析データの生成に失敗しました: \(error.localizedDescription)"
                    completion(.failure(error))
                }
                return
            }


            guard let data = result?.data as? [String: Any] else {
                let parseError = NSError(domain: "Big5AnalysisService", code: 500, userInfo: [NSLocalizedDescriptionKey: "生成データの解析に失敗しました"])
                DispatchQueue.main.async {
                    self?.isLoading = false
                    self?.errorMessage = "生成データの解析に失敗しました"
                    completion(.failure(parseError))
                }
                return
            }


            do {
                // 生成されたデータを解析
                let analysisData = try self?.parseAnalysisData(from: data, personalityKey: personalityKey)
                if let analysisData = analysisData {
                    DispatchQueue.main.async {
                        self?.isLoading = false
                        self?.currentAnalysisData = analysisData
                        // キャッシュに保存
                        self?.cache.cacheAnalysis(key: personalityKey, data: analysisData)
                        completion(.success(analysisData))
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    self?.isLoading = false
                    self?.errorMessage = "生成されたデータの解析に失敗しました"
                    completion(.failure(error))
                }
            }
            }
        }
    }
}

File: flutter/lib/data/services/big5_analysis_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cloud_functions/cloud_functions.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// BIG5分析レベル
enum Big5AnalysisLevel {
  basic, // 20問
  detailed, // 50問
  complete, // 100問
}

/// BIG5分析カテゴリ
enum Big5AnalysisCategory {
  career('career', 'キャリア・仕事'),
  romance('romance', '恋愛・パートナーシップ'),
  stress('stress', 'ストレス対処'),
  communication('communication', 'コミュニケーション'),
  growth('growth', '自己成長');

  final String rawValue;
  final String displayName;

  const Big5AnalysisCategory(this.rawValue, this.displayName);
}

/// BIG5スコア
class Big5Scores {
  final double openness;
  final double conscientiousness;
  final double extraversion;
  final double agreeableness;
  final double neuroticism;

  const Big5Scores({
    required this.openness,
    required this.conscientiousness,
    required this.extraversion,
    required this.agreeableness,
    required this.neuroticism,
  });

  factory Big5Scores.fromMap(Map<String, dynamic> map) {
    return Big5Scores(
      openness: (map['openness'] as num?)?.toDouble() ?? 0,
      conscientiousness: (map['conscientiousness'] as num?)?.toDouble() ?? 0,
      extraversion: (map['extraversion'] as num?)?.toDouble() ?? 0,
      agreeableness: (map['agreeableness'] as num?)?.toDouble() ?? 0,
      neuroticism: (map['neuroticism'] as num?)?.toDouble() ?? 0,
    );
  }

  Map<String, dynamic> toMap() => {
        'openness': openness,
        'conscientiousness': conscientiousness,
        'extraversion': extraversion,
        'agreeableness': agreeableness,
        'neuroticism': neuroticism,
      };
}

/// 詳細分析
class Big5DetailedAnalysis {
  final Big5AnalysisCategory category;
  final String personalityType;
  final String detailedText;
  final List<String> keyPoints;
  final Big5AnalysisLevel analysisLevel;

  const Big5DetailedAnalysis({
    required this.category,
    required this.personalityType,
    required this.detailedText,
    required this.keyPoints,
    required this.analysisLevel,
  });

  factory Big5DetailedAnalysis.fromMap(
    Map<String, dynamic> map,
    Big5AnalysisCategory category,
    Big5AnalysisLevel level,
  ) {
    return Big5DetailedAnalysis(
      category: category,
      personalityType: map['personality_type'] as String? ?? '',
      detailedText: map['detailed_text'] as String? ?? '',
      keyPoints: (map['key_points'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      analysisLevel: level,
    );
  }
}

/// BIG5分析データ
class Big5AnalysisData {
  final String personalityKey;
  final DateTime lastUpdated;
  final Map<Big5AnalysisCategory, Big5DetailedAnalysis>? analysis20;
  final Map<Big5AnalysisCategory, Big5DetailedAnalysis>? analysis50;
  final Map<Big5AnalysisCategory, Big5DetailedAnalysis>? analysis100;

  const Big5AnalysisData({
    required this.personalityKey,
    required this.lastUpdated,
    this.analysis20,
    this.analysis50,
    this.analysis100,
  });

  /// 指定レベルの分析を取得
  Map<Big5AnalysisCategory, Big5DetailedAnalysis>? getAnalysis(
      Big5AnalysisLevel level) {
    switch (level) {
      case Big5AnalysisLevel.basic:
        return analysis20;
      case Big5AnalysisLevel.detailed:
        return analysis50;
      case Big5AnalysisLevel.complete:
        return analysis100;
    }
  }
}

/// BIG5分析キャッシュ
class Big5AnalysisCache {
  static final Big5AnalysisCache shared = Big5AnalysisCache._();
  Big5AnalysisCache._();

  final Map<String, Big5AnalysisData> _cache = {};
  final Map<String, DateTime> _cacheTimestamps = {};
  static const Duration _cacheExpiration = Duration(hours: 24);

  Big5AnalysisData? getCachedAnalysis(String personalityKey) {
    final timestamp = _cacheTimestamps[personalityKey];
    if (timestamp != null) {
      if (DateTime.now().difference(timestamp) > _cacheExpiration) {
        _cache.remove(personalityKey);
        _cacheTimestamps.remove(personalityKey);
        return null;
      }
    }
    return _cache[personalityKey];
  }

  void cacheAnalysis(String key, Big5AnalysisData data) {
    _cache[key] = data;
    _cacheTimestamps[key] = DateTime.now();
  }

  void clearCache() {
    _cache.clear();
    _cacheTimestamps.clear();
  }
}

/// BIG5分析サービス
class Big5AnalysisService {
  final FirebaseFirestore _db;
  final Big5AnalysisCache _cache = Big5AnalysisCache.shared;

  Big5AnalysisService({FirebaseFirestore? firestore})
      : _db = firestore ?? FirebaseFirestore.instance;

  /// パーソナリティキーを生成
  String generatePersonalityKey(Big5Scores scores, String gender) {
    final o = _roundToFiveScale(scores.openness);
    final c = _roundToFiveScale(scores.conscientiousness);
    final e = _roundToFiveScale(scores.extraversion);
    final a = _roundToFiveScale(scores.agreeableness);
    final n = _roundToFiveScale(scores.neuroticism);

    return 'O${o}_C${c}_E${e}_A${a}_N${n}_$gender';
  }

  int _roundToFiveScale(double score) {
    return (score.round()).clamp(1, 5);
  }

  /// 分析データを取得
  Future<Big5AnalysisData> fetchAnalysisData(String personalityKey) async {
    // キャッシュから取得を試行
    final cachedData = _cache.getCachedAnalysis(personalityKey);
    if (cachedData != null) {
      return cachedData;
    }

    final doc =
        await _db.collection('Big5Analysis').doc(personalityKey).get();

    if (!doc.exists || doc.data() == null) {
      throw Exception('解析データがまだ生成されていません');
    }

    final data = doc.data()!;
    final analysisData = _parseAnalysisData(data, personalityKey);
    _cache.cacheAnalysis(personalityKey, analysisData);

    return analysisData;
  }

  /// キャラクターの分析データを取得
  Future<Big5AnalysisData> fetchCharacterAnalysis(
    String characterId,
    String userId,
  ) async {
    final detailsDoc = await _db
        .collection('users')
        .doc(userId)
        .collection('characters')
        .doc(characterId)
        .collection('details')
        .doc('current')
        .get();

    if (!detailsDoc.exists || detailsDoc.data() == null) {
      throw Exception('キャラクター詳細が見つかりませんでした');
    }

    final personalityKey = detailsDoc.data()!['personalityKey'] as String?;
    if (personalityKey == null) {
      throw Exception('personalityKeyが設定されていません');
    }

    return fetchAnalysisData(personalityKey);
  }

  /// 回答数から分析レベルを決定
  Big5AnalysisLevel? determineAnalysisLevel(int answeredCount) {
    if (answeredCount >= 100) {
      return Big5AnalysisLevel.complete;
    } else if (answeredCount >= 50) {
      return Big5AnalysisLevel.detailed;
    } else if (answeredCount >= 20) {
      return Big5AnalysisLevel.basic;
    }
    return null;
  }

  /// レベルで利用可能なカテゴリを取得
  List<Big5AnalysisCategory> getAvailableCategories(Big5AnalysisLevel level) {
    switch (level) {
      case Big5AnalysisLevel.basic:
        return [
          Big5AnalysisCategory.career,
          Big5AnalysisCategory.romance,
          Big5AnalysisCategory.stress,
        ];
      case Big5AnalysisLevel.detailed:
      case Big5AnalysisLevel.complete:
        return Big5AnalysisCategory.values.toList();
    }
  }

  Big5AnalysisData _parseAnalysisData(
    Map<String, dynamic> data,
    String personalityKey,
  ) {
    final lastUpdated =
        (data['last_updated'] as Timestamp?)?.toDate() ?? DateTime.now();

    return Big5AnalysisData(
      personalityKey: personalityKey,
      lastUpdated: lastUpdated,
      analysis20: _parseAnalysisLevel(data, 'analysis_20', Big5AnalysisLevel.basic),
      analysis50: _parseAnalysisLevel(data, 'analysis_50', Big5AnalysisLevel.detailed),
      analysis100: _parseAnalysisLevel(data, 'analysis_100', Big5AnalysisLevel.complete),
    );
  }

  Map<Big5AnalysisCategory, Big5DetailedAnalysis>? _parseAnalysisLevel(
    Map<String, dynamic> data,
    String levelKey,
    Big5AnalysisLevel level,
  ) {
    final levelData = data[levelKey] as Map<String, dynamic>?;
    if (levelData == null) return null;

    final result = <Big5AnalysisCategory, Big5DetailedAnalysis>{};

    for (final category in Big5AnalysisCategory.values) {
      final categoryData = levelData[category.rawValue] as Map<String, dynamic>?;
      if (categoryData != null) {
        result[category] = Big5DetailedAnalysis.fromMap(
          categoryData,
          category,
          level,
        );
      }
    }

    return result.isEmpty ? null : result;
  }

  /// AI分析データを生成（Cloud Functions呼び出し）
  Future<Big5AnalysisData> generateAnalysisData(
    String personalityKey, {
    bool isPremium = false,
  }) async {
    final functions = FirebaseFunctions.instanceFor(region: 'asia-northeast1');
    final callable = functions.httpsCallable('generateBig5Analysis');

    final result = await callable.call({
      'personalityKey': personalityKey,
      'isPremium': isPremium,
    });

    final data = result.data as Map<String, dynamic>;
    final analysisData = _parseAnalysisData(data, personalityKey);
    _cache.cacheAnalysis(personalityKey, analysisData);

    return analysisData;
  }
}

/// プロバイダー
final big5AnalysisServiceProvider = Provider<Big5AnalysisService>((ref) {
  return Big5AnalysisService();
});

final big5AnalysisCacheProvider = Provider<Big5AnalysisCache>((ref) {
  return Big5AnalysisCache.shared;
});
